main/src/library/Fixpoint3/Ram.flix

```diff
@@ -0,0 +1,493 @@
+/*
+ * Copyright 2021 Benjamin Dahse
+ * Copyright 2025 Casper Dalgaard Nielsen
+ *                Adam Yasser Tallouzi
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+mod Fixpoint3.Ast.Ram {
+    // We have not updated these imports to be for Fixpoint3 as we have not changed any of the files
+    use Fixpoint.Ast.Shared.{BoxedDenotation => Denotation, PredSym}
+    use Fixpoint.Ast.Shared.Denotation.{Relational}
+    use Fixpoint.Boxed
+    use Fixpoint.PredSymsOf
+    use Fixpoint.SubstitutePredSym
+
+    /////////////////////////////////////////////////////////////////////////////
+    // RamProgram                                                              //
+    /////////////////////////////////////////////////////////////////////////////
+
+    ///
+    /// A `RamProgram` consists of a statements, the EDB, predicate information,
+    /// and the information about the indexes to be constructed.
+    /// 
+    pub enum RamProgram {
+        case Program(RamStmt, Facts, Predicates, IndexInformation)
+    }
+
+    ///
+    /// `Predicate` represents 1 + the maximal indentifier of the full `RelSym`s of the program and set of `RelSym`
+    ///
+    pub type alias Predicates = (Int64, Set[RelSym])
+
+    ///
+    /// The EDB of a program
+    ///
+    pub type alias Facts = Map[RelSym, Vector[Vector[Boxed]]]
+
+    ///
+    /// A `Search` describes the order in which elements are ordered.
+    /// [2, 0, 1] means that the elements are ordered by the third element,
+    /// then the first and finally the second.
+    ///
+    pub type alias Search = Vector[Int32]
+
+    ///
+    /// `ConstructedIndexes` represents a map from list of indexes created for the RelSym.
+    ///
+    pub type alias ConstructedIndexes = Map[RelSym, Vector[Search]]
+
+    ///
+    /// `IndexPositions` represents a map from RelSym and its place in the vector of ConstructedIndexes
+    /// to its final position in memory.
+    ///
+    pub type alias IndexPositions = Map[(RelSym, Int32), Int32]
```

Comment:
I think its valid to do it as you have, but richer types (aliases or enums) would help readability of all these integer things